From: Simon McVittie <smcv@debian.org>
Date: Tue, 29 Nov 2016 10:17:05 +0000
Subject: test-run: don't assume we are uid 1000

Signed-off-by: Simon McVittie <smcv@debian.org>
---
 tests/test-run.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tests/test-run.sh b/tests/test-run.sh
index fef5367..5b15922 100755
--- a/tests/test-run.sh
+++ b/tests/test-run.sh
@@ -32,7 +32,7 @@ assert_file_has_content () {
 }
 
 FUSE_DIR=
-for mp in `cat /proc/self/mounts | grep " fuse[. ]" | grep user_id=1000 | awk '{print $2}'`; do
+for mp in `cat /proc/self/mounts | grep " fuse[. ]" | grep user_id=`id -u` | awk '{print $2}'`; do
     if test -d $mp; then
         echo Using $mp as test fuse mount
         FUSE_DIR=$mp
